### DISCUSSION TITLE
Getting a No such shell function `zle-line-init' in Terminal

### DISCUSSION THREAD

**Discussion by ch-lee** [JanDeDobbeleer/oh-my-posh#3462] [Troubleshoot] [answered]
Hi there,

I'm getting a `No such shell function 'zle-line-init'` in the terminal (I'm using iTerm2), just underneath the prompt. See screenshot:

<img width="571" alt="Screenshot 2023-02-06 at 21 33 25" src="https://user-images.githubusercontent.com/3100303/217092876-39595948-1551-40b7-be3f-0174b8c4c357.png">

I'm using the theme `jandedobbeleer.omp.json` - no changes made. Just a straight clone.

I'm getting this each time I press the enter key.

I noticed that I don't have this error when I just use the default theme - see screenshot below:

<img width="568" alt="Screenshot 2023-02-06 at 21 34 50" src="https://user-images.githubusercontent.com/3100303/217093124-5250e000-c8c6-4143-a2cb-0bc8582f523d.png">

But, when I press enter key or any command, I get this:

<img width="563" alt="Screenshot 2023-02-06 at 21 36 16" src="https://user-images.githubusercontent.com/3100303/217093338-f0a7f19e-58af-45fc-a545-baa3c0a49d9a.png">

I believe this is the [Transient Prompt](https://ohmyposh.dev/docs/configuration/transient)? 

I have seen the [how to disable Transient prompt ](https://github.com/JanDeDobbeleer/oh-my-posh/discussions/1527) and I don't have the `transient_prompt` json element set in the `jandedobbeleer.omp.json`

Inside my `.zshrc` file:

```
if [ $TERM_PROGRAM != "Apple_Terminal" ]; then
  eval "$(oh-my-posh init zsh)"
fi

# eval "$(oh-my-posh init zsh --config ~/.jandedobbeleer.omp.json)"
```

How can I stop the `No such shell function 'zle-line-init'` from appearing? I presume I am missing something? 

Any help appreciated!

**Comment by JanDeDobbeleer** [comment]
@ch-lee `eval "$(oh-my-posh init zsh)"` is what gets run as `eval "$(oh-my-posh init zsh --config ~/.jandedobbeleer.omp.json)"` is commented out and the default theme (embedded) does have a transient prompt.

That said, your zsh is either outdated OR something is hijacking/removing those native zsh functions (see ohmyzsh or similar).

**Comment by ch-lee** [comment]
Thanks for the quick reply.

Sorry, I was commenting out `eval "$(oh-my-posh init zsh --config ~/.jandedobbeleer.omp.json)"` in my `.zshrc` file to switch between the default theme and manually setting a theme.

I thought my zsh was out of date as well and did a `brew reinstall zsh`.

Current version installed now is: `zsh 5.9 (arm-apple-darwin22.1.0)` 

It's very strange. I don't have ohmyzsh installed either.

**Reply by JanDeDobbeleer** [reply]
@ch-lee hmm, this is a bit strange tbh. That function should be part of zsh' builtin ZLE functionality. I'll have a look tomorrow.

**Reply by ch-lee** [reply]
Yeah, it super strange. I'm not too familiar with ZLE functionality. Been trying to re-install lots of things at the moment. Lots of googling as well.  Will keep trying. Let me know if you have any ideas tomorrow. Thanks

**Reply by mahbam42** [reply]
Found this thread because I did the same thing... tried to comment out the first eval line to test a theme and got the same error. Removing the commented line resolved it immediately

**Comment by Tuttu** [comment]
Hello, 

I'm having the same issue with both zsh and oy-my-posh updated to the latest version.
I'm using a theme that doesn't have Transient Prompt for what it matters.

**Comment by JanDeDobbeleer** [comment]
@Tuttu is your zsh version up-to-date?

**Reply by Tuttu** [reply]
Yep. I updated it to the latest version `zsh 5.9 (arm-apple-darwin21.3.0)` before answering here. :)

**Reply by JanDeDobbeleer** [reply]
@Tuttu nice, are you using oh my zsh or any other plugin?

**Reply by Tuttu** [reply]
@JanDeDobbeleer Nope. It's a fresh macOS install and I have only installed brew to be able to install oh-my-bosh. There shouldn't be anything else unless my IT team is installing something in the background which I doubt.

**Reply by Tuttu** [reply]
One thing to note: if I'm using the `exec zsh` command in my terminal window, zsh will reload without that 'error' message being displayed.

**Reply by JanDeDobbeleer** [reply]
@Tuttu that implies something isn't loaded and that could be related to your terminal setup and how it invokes zsh.

**Reply by Tuttu** [reply]
Well. I will check my installation and see if something is wrong with it.

**Comment by GaryBoone** [comment, marked as answer]
I just had this error, `No such shell function 'zle-line-init'` and found that it was because I was calling `eval "$(oh-my-posh init ..."` _twice_ in `.zshrc`, once to load Oh My Posh, then another to load the theme. Using just one worked:
```
eval "$(oh-my-posh init zsh --config <path>/<theme_name>.omp.json)"
```

**Reply by Tuttu** [reply]
It worked perfectly for me as well. Thanks.
I guess it might be worth to add that to the FAQ. :)

**Reply by JanDeDobbeleer** [reply]
@Tuttu @GaryBoone I'm interested in how you came to the conclusion to add that line rather adjust the existing one?

**Reply by Tuttu** [reply]
@JanDeDobbeleer On my hand, that's because I'm all new with zsh and I've been stupidly following the different steps of the documentation to setup oh-my-posh.
First, add the init line without the custom them to start oh-my-posh with the default theme. Then, following the wiki on how to use custom themes, I added a new line rather than changing the first one.

**Reply by JanDeDobbeleer** [reply]
@Tuttu I'll give the wording a go.

**Reply by JanDeDobbeleer** [reply]
https://github.com/JanDeDobbeleer/oh-my-posh/pull/3543

**Reply by nbennett25** [reply]
Hey @JanDeDobbeleer  - Still seeing this with version `16.7.1`, and I only have one eval line in my `.zshrc` file : /  

```
  grep posh .zshrc
eval "$(oh-my-posh init zsh --config ~/avit.omp.json)"
~ 
  
No such shell function `zle-line-init'
```

zsh version:
`zsh 5.8.1 (x86_64-apple-darwin21.0)`

**Reply by JanDeDobbeleer** [reply]
@nbennett25 that's either a zsh version issue (maybe outdated), or something else outside of our control.

**Reply by lucabezerra** [reply]
My 2 cents: In my case, I added the `eval` command to my `.zshrc` only once, yet I was also getting this error/warning and couldn't figure out why. Later I realized I ran in the same terminal the command to test a theme on the spot (`eval "$(oh-my-posh init zsh --config $(brew --prefix oh-my-posh)/themes/<some_other_theme>.omp.json)"`), somewhat similar to what @Tuttu did, although less permanent. Turns out I just needed to restart the terminal 🤦🏻  Hope it helps someone, haha.

**Reply by bacf5** [reply]
Worked for me. Thanks

**Comment by ai-mine-max** [comment]
I see the following error each time I open my terminal or bash:
```
autoload: command not found
add-zsh-hook: command not found
add-zsh-hook: command not found
Command 'zle' not found, did you mean:
  command 'mle' from snap mle-editor (1.5.0)
  command 'ale' from deb ale (0.9.0.3-5)
  command 'zile' from deb zile (2.6.2-2)
  command 'gle' from deb gle-graphics (4.2.5-9)
  command 'le' from deb le (1.16.8-0.1)
  command 'mle' from deb mle (1.4.3-2)
  command 'zl' from deb zenlisp (2013.11.22-3)
```
after adding 
```
eval "$(oh-my-posh init zsh --config ~/<ttheme>.omp.json)" 
```
to my  ```~/.zshrc```file.

**Reply by JanDeDobbeleer** [reply]
@ai-mine-max this is a zsh configuration/installation issue. See [this thread](https://github.com/ohmyzsh/ohmyzsh/issues/6405) for pointers.
